From d9e6c5b30f1974a32e438660413fee7c0e785be5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Edvinas=20Stun=C5=BE=C4=97nas?= <edvinas@8devices.com>
Date: Fri, 14 Mar 2025 10:29:41 +0000
Subject: [PATCH] wifi: ath10k: snoc: add MAC address override module param
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Specifying MAC address in DTS is not feasible in most cases, also not
all platforms supports or are eligible to keep and retrieve HW address
from nvmem. Module parameter seems to be good alternative to specify
MAC address and control it from user space.

Signed-off-by: Edvinas Stunžėnas <edvinas@8devices.com>
Upstream-Status: Inappropriate [embedded specific]
---
 drivers/net/wireless/ath/ath10k/snoc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/net/wireless/ath/ath10k/snoc.c b/drivers/net/wireless/ath/ath10k/snoc.c
index 2c39bad7ebfb..89de81ab0d6c 100644
--- a/drivers/net/wireless/ath/ath10k/snoc.c
+++ b/drivers/net/wireless/ath/ath10k/snoc.c
@@ -7,6 +7,7 @@
 #include <linux/clk.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <linux/etherdevice.h>
 #include <linux/of.h>
 #include <linux/of_device.h>
 #include <linux/platform_device.h>
@@ -471,6 +472,10 @@ static struct ce_service_to_pipe target_service_to_ce_map_wlan[] = {
 	},
 };
 
+static char *mac_addr;
+module_param(mac_addr, charp, 0644);
+MODULE_PARM_DESC(mac_addr, "Network device MAC address");
+
 static void ath10k_snoc_write32(struct ath10k *ar, u32 offset, u32 value)
 {
 	struct ath10k_snoc *ar_snoc = ath10k_snoc_priv(ar);
@@ -1713,6 +1718,7 @@ static int ath10k_snoc_probe(struct platform_device *pdev)
 	struct ath10k_snoc *ar_snoc;
 	struct device *dev;
 	struct ath10k *ar;
+	u8 maddr[ETH_ALEN];
 	u32 msa_size;
 	int ret;
 	u32 i;
@@ -1745,6 +1751,9 @@ static int ath10k_snoc_probe(struct platform_device *pdev)
 	ar->ce_priv = &ar_snoc->ce;
 	msa_size = drv_data->msa_size;
 
+	if (mac_addr && mac_pton(mac_addr, maddr) && is_valid_ether_addr(maddr))
+		memcpy(ar->mac_addr, maddr, sizeof(ar->mac_addr));
+
 	ath10k_snoc_quirks_init(ar);
 
 	ret = ath10k_snoc_resource_init(ar);
